Configuration files

breed808 opened the pull request:
Initial implementation for configuration file support in windows_exporter.
Currently, the exporter supports reading from multiple YAML configuration files specified with the argument `--config.files=file1.yml,file2.yml`.
Values defined in these configuration files have a higher priority than values defined via command line arguments.

This will need testing and documentation once we're happy with the method/syntax of reading the files, and the config file structure.
Once merged, this will resolve #600.

carlpett reviewed (commented):
Really nice! Some comments inlined, but the two larger questions for me are around how we'll map yaml to flags.
First, the dot notation in flag names sort of implies nesting, but I don't see that that would happen as is, right?
Second, if we do want to support nesting, we need to decide how this should overlap with the multiple files support. Merging 
```yaml
collectors:
  mssql:
    classes-enabled:
    - bufman
    - databases
```

with eg
```yaml
collectors:
  mssql: {}
  # or mssql: null
```

Should these empty out the first list of collectors? 

I'm not sure it's wise to try to solve the combination of the two features in the first implementation? There's a lot of edge cases.

breed808 commented on exporter.go:
```diff
@@ -268,36 +284,62 @@ func initWbem() {
+		app         = kingpin.New("windows_exporter", "")
+		configFiles = app.Flag(
+			"config.files",
+			"Comma-separated list of YAML configuration file(s) to use. Values set in these file(s) will override flags.",
```
Agreed on the priority, I'll amend this to set the flags as higher priority.
Would support for multiple configurations files be desirable, or would a single config file be acceptable initially?

carlpett replied:
I think a single file is a good start to iron out how we want the basic things to work and lets us avoid a lot of these tricky questions for now :) I'd also guess that it would be sufficient for most users.

breed808 commented on config.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// Resolver represents a configuration file resolver for kingpin.
+type Resolver struct {
```
Good point, I'll move it to an internal `config` sub-package.

breed808 commented:
I've moved `config.go` to a separate file, and removed support for multiple config files.
Priorities of CLI flags vs config file values needs to be resolved, as does the format.

Currently the following is valid:

```yaml
collectors.enabled: cpu,cs,net
log.level: warn
```
Which is not desirable. Parser should support this format:
```yaml
collectors:
  enabled: cpu,cs,net
log:
  level: warn
```

breed808 commented:
I've looked into the structure of the configuration file further, and it looks like the code I shamelessly lifted from `amtool` expects a flat yaml file, unmarshalling into a `map[string]string`. For windows_exporter, this would look like:

```yaml
collectors.enabled: cpu,cs,net
log.level: warn
```

A nested config file could be implemented, but I think we'd need to define a config struct to unmarshal the file into, which would result in maintaining configuration options in two areas (CLI flags and config file structure). The alternative would be heavy use of `interface{}` and manual type checking of configuration input.

It looks like kingpin supports reading flags from a specified file, though this wouldn't be in the expected YAML format:
```
--enabled.collectors=cpu,cs,net
--log.level=warn
```

I've looked at `node_exporter` and it appears most of their functionality is exposed via CLI flags only, with some TLS settings exposed via YAML config only.
Are there any known examples of kingpin flags being read from CLI flags **and** config files?

carlpett commented:
Good points - totally agree that we don't want to maintain that mapping manually. Looking at the code you've written so far, I _think_ we could get away with unmarshalling into a `map[string]interface{}`, and then continuing the shameless stealing and doing some light adaptation of [this flattening code](https://github.com/carlpett/terraform-provider-sops/blob/master/sops/flatten.go). I think we just need to change the slice handling to comma-separate instead of adding indices to the key. What do you think, would that work?

breed808 commented:
That flattening code is exactly what was needed. I've tested it on a nested configuration file and it's working!

carlpett reviewed (approved):
Really neat!
